### PR TITLE
Auto-stock for stocking input bus

### PR DESCRIPTION
**Pull request by MadMan310**
This special mode for the stocking input bus will grab the first 16 items in the ME system it is connected to, and add them to it's list of item to stock.
The list is updated every 5 seconds while it is in this mode.
Toggle using right click + screwdriver.

This is meant for automation setups that use stocking input buses to supply machines.
A combination of a subnetwork + stocking buses is currently the most efficient method of routing items, especially for autocrafting setups.
However, this forces the player to update every stocking input bus they have on each machine if they ever wanted to add a new pattern.
This mode is meant to compliment those setups, and is primarily meant for use on small subnetworks.

### CODE REVIEW

**Comment by LewisSaber** [comment]
Is it possible to get some copy-paster? Like glease did there https://github.com/GTNewHorizons/GigaGramFab/pull/6

**Comment by MadMan310** [comment]
> Is it possible to get some copy-paster? Like glease did there [GTNewHorizons/GigaGramFab#6](https://github.com/GTNewHorizons/GigaGramFab/pull/6)
Looking at it, this looks relatively straightforward to implement.

**Comment by LewisSaber** [comment]
It loads in wrong order
Original
image
Copy
image

**Comment by Dream-Master** [comment]
@MadMan310

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 1]
```diff
@@ -135,11 +166,25 @@ public void saveNBTData(NBTTagCompound aNBT) {
...
 
+    private void setAutoPullItemList(boolean pullItemList) {
+        autoPullItemList = pullItemList;
+        if (!autoPullItemList) {
+            for (int i = 0; i < SLOT_COUNT; i++) {
+                mInventory[i] = null;
+            }
+        } else {
+            refreshItemList();
+            updateAllInformationSlots();
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:184]: `updateAllInformationSlots()` should be called on both cases

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 2]
```diff
@@ -135,11 +166,25 @@ public void saveNBTData(NBTTagCompound aNBT) {
...
         aNBT.setIntArray("sizes", sizes);
+        aNBT.setBoolean("autoStock", autoPullItemList);
+        aNBT.setInteger("minAutoPullStackSize", minAutoPullStackSize);
         if (GregTech_API.mAE2) {
             gridProxy.writeToNBT(aNBT);
         }
     }
 
+    private void setAutoPullItemList(boolean pullItemList) {
+        autoPullItemList = pullItemList;
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:177]: Check for not being on client side is needed I guess

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 3]
```diff
@@ -187,11 +234,68 @@ public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex
...
-    public void onScrewdriverRightClick(byte aSide, EntityPlayer aPlayer, float aX, float aY, float aZ) {}
+    public void onScrewdriverRightClick(byte aSide, EntityPlayer aPlayer, float aX, float aY, float aZ) {
+        setAutoPullItemList(!autoPullItemList);
+        GT_Utility.sendChatToPlayer(aPlayer, "Automatic Item Pull " + autoPullItemList);
+    }
 
     @Override
     public void updateSlots() {}
 
+    public boolean onRightclick(IGregTechTileEntity aBaseMetaTileEntity, EntityPlayer aPlayer, byte aSide, float aX,
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:245]: missing `@Override`

**Comment by Dream-Master** [comment]
@MadMan310

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 4]
```diff
@@ -363,12 +500,71 @@ private boolean containsSuchStack(ItemStack tStack) {
                                 .build().setPos(97, 9))
                 .widget(
                         new DrawableWidget().setDrawable(GT_UITextures.PICTURE_ARROW_DOUBLE).setPos(82, 40)
-                                .setSize(12, 12));
+                                .setSize(12, 12))
+                .widget(new ButtonWidget().setOnClick((clickData, widget) -> {
+                    if (clickData.mouseButton == 0) {
+                        setAutoPullItemList(!autoPullItemList);
+                    } else if (clickData.mouseButton == 1 && !widget.isClient()) {
+                        widget.getContext().openSyncedWindow(10);
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:508]: Use `CONFIG_WINDOW_ID`

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 5]
```diff
@@ -179,6 +180,7 @@ private void setAutoPullItemList(boolean pullItemList) {
             for (int i = 0; i < SLOT_COUNT; i++) {
                 mInventory[i] = null;
             }
+            updateAllInformationSlots();
         } else {
             refreshItemList();
             updateAllInformationSlots();
```
Comment (miozune) [inline src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:186]: Now you can move these two calls to outside of if-else clause

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 6]
```diff
@@ -187,11 +234,68 @@ public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex
...
+    public boolean onRightclick(IGregTechTileEntity aBaseMetaTileEntity, EntityPlayer aPlayer, byte aSide, float aX,
+            float aY, float aZ) {
+        if (!(aPlayer instanceof EntityPlayerMP))
+            return super.onRightclick(aBaseMetaTileEntity, aPlayer, aSide, aX, aY, aZ);
+        ItemStack dataStick = aPlayer.inventory.getCurrentItem();
+        if (!ItemList.Tool_DataStick.isStackEqual(dataStick, true, true))
+            return super.onRightclick(aBaseMetaTileEntity, aPlayer, aSide, aX, aY, aZ);
+
+        if (!dataStick.hasTagCompound() || !"stockingBus".equals(dataStick.stackTagCompound.getString("type")))
+            return true;
```
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:257]: unresolved

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java [hunk 7]
```diff
@@ -187,11 +234,68 @@ public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex
...
+                this.mInventory[i] = GT_Utility.loadItem(stockingItems.getCompoundTagAt(i));
+            }
+        }
+        setInventorySlotContents(getCircuitSlot(), circuit);
+        aPlayer.addChatMessage(new ChatComponentText("Loaded Config From Data Stick"));
+        return true;
+    }
+
+    @Override
+    public void onLeftclick(IGregTechTileEntity aBaseMetaTileEntity, EntityPlayer aPlayer) {
```
[the remark this answers is not in the document]
Comment (miozune) [reply src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:277]: unresolved

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java (hunk 2 above)
Comment (miozune) [reply src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:178]: unresolved

**Comment by Dream-Master** [comment]
@MadMan310

File: src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (MadMan310) [reply src/main/java/gregtech/common/tileentities/machines/GT_MetaTileEntity_Hatch_InputBus_ME.java:277]: The `onRightClick` method is not called at all if the player is sneaking, so it is impossible to add this functionality.
This is why I am having it save the configuration on left click

**Review by miozune — APPROVED** [review]

**Pull request merged**